qutip/core/cy/qobjevo.pyx

```diff
@@ -0,0 +1,1114 @@
+                coeff_elements.append(element)
+            else:
+                func_elements.append(element)
+
+        cleaned_elements = []
+        if len(cte_elements) >= 2:
+            # Multiple constant parts
+            cleaned_elements.append(_CteElement(sum(element.qobj for element in cte_elements)))
+        else:
+            cleaned_elements += cte_elements
+
+        coeff_elements = self._compress_merge_qobj(coeff_elements)
+        cleaned_elements += coeff_elements + func_elements
+
+        self.elements = cleaned_elements
+
+    ###########################################################################
+    # properties                                                              #
+    ###########################################################################
+    @property
+    def num_obj(self):
+        """Number of parts composing the system"""
+        return len(self.elements)
+
+    @property
+    def const(self):
+        """Does the system change depending on `t`"""
+        return not any(type(element) is not _CteElement for element in self.elements)
+
+    @property
+    def isoper(self):
+        """Indicates if the system represents an operator."""
+        # TODO: isoper should be part of dims
+        return type_from_dims(self.dims) == "oper"
+
+    @property
+    def issuper(self):
+        """Indicates if the system represents a superoperator."""
+        # TODO: issuper should/will be part of dims
+        # remove self._issuper then
+        if self._issuper == -1:
+            self._issuper = type_from_dims(self.dims) == "super"
+        return self._issuper
+
+    ###########################################################################
+    # operation methods                                                       #
+    ###########################################################################
+    def expect(QobjEvo self, t, state):
+        """
+        Expectation value of this operator at time `t` with the state.
+
+        Parameters
+        ----------
+        t : float
+            Time of the operator to apply.
+        state : Qobj
+            right matrix of the product
+
+        Returns
+        -------
+        expect : float or complex
+            `state.adjoint() @ self @ state` if `state` is a ket.
+            `trace(self @ matrix)` is `state` is an operator or operator-ket.
+        """
+        if not isinstance(t, (int, float)):
+            raise TypeError("the time need to be a real scalar")
+        if not isinstance(state, Qobj):
+            raise TypeError("A Qobj state is expected")
+        if not (
+            (self.dims[1] == state.dims[0]) or
+            (self.issuper and self.dims[1] == state.dims)
+        ):
+            raise ValueError("incompatible dimensions " + str(self.dims) +
+                             ", " + str(state.dims))
+        out = self.expect_data(t, state.data)
+        if fabs(out.imag) < 1e-15:
+            return out.real
```

Comment:
QobjEvo cannot detect their hermiticity without looking into Coefficient... The options I see are:
- Have an hermitian flag manually set at creation (and trust it blindly).
- `QobjEvo.expect` always return a complex. 
- Check for if the expectation value is effectively real. Here I should have used a relative test: `abs(out.imag/out.real) < 1e-15`, still not ideal since `0` will always be complex.
- Compute the hermiticity of op(t) each time. (But this would be too slow)

Reply:
Yeah, I get all the problems with determining a hermicity for `QobjEvo`. I think anything we do to make it "knowable" will be pretty useless at best, and endlessly frustrating for users at worst - it'll be hard to do right even in things like `mesolve`.

I'm not sure I can see how any of the `Qobj` properties can work on `QobjEvo`, to be honest (e.g. `isherm`, `istp`, `isunitary` etc).  We probably shouldn't define them.

For now, I guess my suggestion is to a) lookup the tolerance in a settings object somewhere and b) swap it to a relative tolerance (with a little care when the real part is zero).